File: 2025/05 May/10 Longest Subarray with Majority Greater than K/main.cpp
// Link: https://www.geeksforgeeks.org/problems/longest-subarray-with-majority-greater-than-k/1

#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    int longestSubarray(vector<int> &arr, int k)
    {
        int n = arr.size();
        unordered_map<int, int> prefIdx;
        int sum = 0, res = 0;

        for (int i = 0; i < n; i++)
        {

            sum += (arr[i] > k ? 1 : -1);

            if (prefIdx.find(sum) == prefIdx.end())
                prefIdx[sum] = i;
        }

        if (prefIdx.find(-n) != prefIdx.end())
            return 0;

        prefIdx[-n] = n;

        for (int i = -n + 1; i <= n; i++)
        {
            if (prefIdx.find(i) == prefIdx.end())
                prefIdx[i] = prefIdx[i - 1];
            else
                prefIdx[i] = min(prefIdx[i], prefIdx[i - 1]);
        }

        sum = 0;
        for (int i = 0; i < n; i++)
        {
            sum += (arr[i] > k ? 1 : -1);
            if (sum > 0)
                res = i + 1;
            else
                res = max(res, i - prefIdx[sum - 1]);
        }
        return res;
    }
};

int main()
{
    Solution obj;
    vector<int> arr = {1, 2, 3, 4, 1};
    int k = 2;
    cout << obj.longestSubarray(arr, k) << endl;

    arr = {6, 5, 3, 4};
    k = 2;
    cout << obj.longestSubarray(arr, k) << endl;

    return 0;
}

File: 2025/05 May/10 Longest Subarray with Majority Greater than K/main.py
# Link: https://www.geeksforgeeks.org/problems/longest-subarray-with-majority-greater-than-k/1
class Solution:
    def longestSubarray(self, arr, k):
        n = len(arr)
        prefIdx = {}
        sum = 0
        res = 0

        for i in range(n):
            sum += 1 if arr[i] > k else -1

            if sum not in prefIdx:
                prefIdx[sum] = i

        if -n in prefIdx:
            return 0
        prefIdx[-n] = n

        for i in range(-n + 1, n + 1):
            if i not in prefIdx:
                prefIdx[i] = prefIdx[i - 1]
            else:
                prefIdx[i] = min(prefIdx[i], prefIdx[i - 1])

        sum = 0
        for i in range(n):
            sum += 1 if arr[i] > k else -1
            if sum > 0:
                res = i + 1
            else:
                res = max(res, i - prefIdx[sum - 1])

        return res


if __name__ == "__main__":
    obj = Solution()
    arr = [1, 2, 3, 4, 1]
    k = 2
    print(obj.longestSubarray(arr, k))

    arr = [6, 5, 3, 4]
    k = 2
    print(obj.longestSubarray(arr, k))
